adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/lwm2m/LwM2MResourceDirectory.java

```diff
@@ -0,0 +1,694 @@
+            observationServiceImpl.setNonSecureEndpoint(ep);
+        });
+
+        lwm2mRequestSender = new CaliforniumLwM2mRequestSender(
+                getAdapter().getSecureEndpoint(),
+                getAdapter().getInsecureEndpoint(),
+                observationServiceImpl,
+                modelProvider,
+                new DefaultLwM2mNodeEncoder(),
+                nodeDecoder);
+
+        observationServiceImpl.addListener(new ObservationListener() {
+
+            @Override
+            public void onResponse(
+                    final Observation observation,
+                    final Registration registration,
+                    final ObserveResponse response) {
+
+                getAdapter().runOnContext(go -> {
+                    handleNotification(registration, response);
+                });
+            }
+
+            @Override
+            public void onError(
+                    final Observation observation,
+                    final Registration registration,
+                    final Exception error) {
+            }
+
+            @Override
+            public void newObservation(final Observation observation, final Registration registration) {
+            }
+
+            @Override
+            public void cancelled(final Observation observation) {
+            }
+        });
+
+        observationService = observationServiceImpl;
+        lwm2mRegistrationStore.start();
+
+        return Future.succeededFuture();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Future<Void> stop() {
+        if (lwm2mRegistrationStore != null) {
+            lwm2mRegistrationStore.stop();
+        }
+        return Future.succeededFuture();
+    }
+
+    private String getRegistrationId(final Device device) {
+        return String.format("%s:%s", device.getTenantId(), device.getDeviceId());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Future<CoapContext> createCoapContextForPost(final CoapExchange coapExchange, final Span span) {
+        return getAuthenticatedDevice(coapExchange)
+                .map(device -> CoapContext.fromRequest(coapExchange, device, device, null, span));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected Future<CoapContext> createCoapContextForDelete(final CoapExchange coapExchange, final Span span) {
+        return getAuthenticatedDevice(coapExchange)
+                .map(device -> CoapContext.fromRequest(coapExchange, device, device, null, span));
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Handles a request to create or update a device's registration.
+     */
+    @Override
+    protected Future<?> handlePost(final CoapContext ctx) {
+
+        final Request request = ctx.getExchange().advanced().getRequest();
+        LOG.debug("processing POST request: {}", request);
+
+        // we only support confirmable messages
+        if (!ctx.isConfirmable()) {
```

Comment:
Is it possible to add some diagnose text to the response?
e.g. "lwm2m doesn't support NON requests!".